### PR TITLE
planner, executor: supports select statement with AS OF

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

Ref: https://github.com/pingcap/tidb/issues/24557
Problem Summary:

Implement the syntax `Select .... FROM xxx AS OF timestamp xxx` with stale read.

### What is changed and how it works?

the planner should visit the `as of` clause and extract the tso.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- make select statements support stale read

### CODE REVIEW

File: planner/core/logical_plan_builder.go
Comment (JmPotato) [inline planner/core/logical_plan_builder.go:2483]: In #24328, we assume that the type of `tidb_bounded_staleness(a, b)`'s result is `Int` TSO, considering the use of `as of timestamp TIMESTAMP()`, should we unify them to `TIMESTAMP` type?

Comment (djshow832) [inline planner/core/logical_plan_builder.go:2483]: I just tried `select * from t, b;`  and it panics: both `first` and `val` are nil.

File: planner/core/logical_plan_builder.go
Comment (djshow832) [inline planner/core/logical_plan_builder.go:2480]: `ts` is not clear enough for users.

File: planner/core/logical_plan_builder.go
Comment (nolouch) [reply planner/core/logical_plan_builder.go]: yes, better unify them to timestamp type. I have used `ConvertTo` to do that.

**Comment by nolouch** [comment]
PTAL @winoros @djshow832 @Yisaer @JmPotato

File: planner/optimize.go
Comment (winoros) [inline planner/optimize.go:347]: Could we explain stale read here or just post a link here?

File: planner/core/logical_plan_builder.go
Comment (winoros) [inline planner/core/logical_plan_builder.go:2486]: This check should be a global check?
`select ... from (select ... from t as of ...) tmp_1, (select ... from t2 as of ...) tmp_2 ...`
We should guarantee that the AsOfs in the two sub selects use the same ts?

File: planner/core/logical_plan_builder.go
Comment (winoros) [inline planner/core/logical_plan_builder.go:2458]: We could add comments for when we would throw errors out.

File: planner/core/logical_plan_builder.go
Comment (winoros) [reply planner/core/logical_plan_builder.go:2486]: Also for another case, `select ... from (select ... from t as of ...) tmp_1 as of xxx` is valid or not?

File: executor/stale_txn_test.go
Comment (xhebox) [inline executor/stale_txn_test.go:121]: `mockStalenessTxnSchemaVer` is removed in the compability PR.

File: executor/stale_txn_test.go
Comment (xhebox) [inline executor/stale_txn_test.go:195]: Duplicated `use test`.

File: executor/stale_txn_test.go
Comment (xhebox) [inline executor/stale_txn_test.go:205]: Maybe exchange two branches? Then `IsNil` branch will be closer to the examination below.

File: planner/core/logical_plan_builder.go
Comment (xhebox) [reply planner/core/logical_plan_builder.go:2486]: Yes, it should be a global check. If `startTS` is same, then it is valid. Because we bind only one txn with one `startTS`.

It is kinda meaningless talking about `tmp_1 as of xxx`, the temporary created from sub select stmt does not have a history.

File: planner/core/logical_plan_builder.go
Comment (xhebox) [inline planner/core/logical_plan_builder.go:2446]: I think we could add simple `hasCurrentTS` boolean flag in the resolver. Then the condition is like `hasCurrentTS == true` or `len(tsValues) > 0`, but not both.

**Review by xhebox — COMMENTED** [review]
LGTM, expect that I am not sure if handles nested select stmt well. I doubt that it needs to be recursively called.

File: planner/optimize.go
Comment (JmPotato) [inline planner/optimize.go:358]: Better use `oracle.GoTimeToTS`.

**Comment by xhebox** [comment]
Maybe inject the login here, https://github.com/pingcap/tidb/blob/master/planner/core/preprocess.go#L126. I think the current code does not handle nested cases.

**Comment by nolouch** [comment]
@xhebox I found after merged https://github.com/pingcap/tidb/pull/24285. the test case will be failed because the table does not exist.

**Comment by xhebox** [comment]
@nolouch One obvious reason is that, now staleread is a real history read... So you can not fake timestamp anymore. A read with `-20s` will definitely fail because table has not been created 20s before.

**Comment by xhebox** [comment]
@nolouch check https://github.com/xhebox/tidb/tree/stale_1, or https://github.com/xhebox/tidb/commit/03e84552337c1db16c3a6c2ea985b98729c13073, if you think it is fine. I'll push to this PR directly.

Sadly, we can not support infoschema in the planner due to some cycle import problem on the tests. So here is a trick without modifying any API.

**Comment by nolouch** [comment]
It's ok to me @xhebox .

**Comment by xhebox** [comment]
Done. Waiting for CI.

File: executor/stale_txn_test.go
Comment (Yisaer) [inline executor/stale_txn_test.go:113]: I think we should test this case:
``` 
begin
select .... as of timestamp (expecting error)
```

Comment (nolouch) [reply executor/stale_txn_test.go:113]: Yes, let us handle it in another PR.

File: planner/core/preprocess.go
Comment (JmPotato) [inline planner/core/preprocess.go:1370]: I think we can reuse the `calculateTsExpr` function defined in [planbuilder.go#L2345]( https://github.com/pingcap/tidb/blob/master/planner/core/planbuilder.go#L2345) to do the same thing. A little modification to that function signature is needed though.

Comment (xhebox) [reply planner/core/preprocess.go:1370]: Yeah, but that function must be extracted somewhere.

File: executor/adapter.go
Comment (JmPotato) [inline executor/adapter.go:325]: I think the timestamp that `oracle.ExtractPhysical` returns is in millisecond, and the tolerance is 2 seconds if I remember correctly.

**Comment by nolouch** [comment]
PTAL @winoros  @JmPotato @Yisaer

File: executor/stale_txn_test.go
Comment (JmPotato) [inline executor/stale_txn_test.go:418]: This may be unnecessary after #24820 is merged.

File: go.mod
Comment (JmPotato) [inline go.mod:84]: Is the `go.mod` and `go.sum` change in expect?

File: executor/stale_txn_test.go
Comment (xhebox) [reply executor/stale_txn_test.go:418]: Yes, it is.

File: executor/adapter.go
Comment (nolouch) [reply executor/adapter.go:325]: Great catch.

File: go.mod
Comment (nolouch) [reply go.mod:84]: Yes, I do it with `rm go.sum && go mod tidy` is same as now.

**Comment by xhebox** [comment]
Maybe there are stable test failures. I'll look into it.

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

**Comment by xhebox** [comment]
@nolouch Let me handle the fix. I have figured out why.

**Comment by xhebox** [comment]
I should have tried 880541b fix but it didn't work for me at the time. Why it works now..?

**Comment by nolouch** [comment]
@xhebox I think the schema is changed, but we use `Preprocess schema` again is the problem. But I don’t know why there is no work at your place.

**Comment by nolouch** [comment]
PTAL @JmPotato @Yisaer

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

File: planner/core/preprocess.go
Comment (nolouch) [reply planner/core/preprocess.go]: It will fail with `can not set different time in the as of` in the `TimestampExactRead1` case.

**Review by Yisaer — APPROVED** [review]
/lgtm

**Review by JmPotato — APPROVED** [review]

File: planner/optimize.go
Comment (winoros) [inline planner/optimize.go:105]: This blank is used to split code blocks, better to reserve:joy:

File: planner/core/errors.go
Comment (winoros) [inline planner/core/errors.go:100]: You can add a new error no in the package `errorno`'s `TiDB self-defined errors.` part. The code may be used for some analysis.

File: planner/core/common_plans.go
Comment (winoros) [inline planner/core/common_plans.go:268]: The fix for `prepare` will be given soon?

File: executor/prepared.go
Comment (winoros) [inline executor/prepared.go:179]: Actually the `infoschema` should keep unchanged afte the change?

File: planner/core/common_plans.go
Comment (xhebox) [reply planner/core/common_plans.go:268]: Or ban if it can not be done very soon.

File: executor/prepared.go
Comment (xhebox) [reply executor/prepared.go:179]: Stale reads mean to read things at a specific time. So if we will support prepare stale select eventually, it should be unchanged.

File: planner/core/errors.go
Comment (nolouch) [reply planner/core/errors.go:100]: We may support it in the future, so we don't need to occupy a place in errorno?

File: planner/core/common_plans.go
Comment (nolouch) [reply planner/core/common_plans.go:268]: I create a issue for this: https://github.com/pingcap/tidb/issues/24932

**Review by winoros — APPROVED** [review]

File: planner/core/logical_plan_test.go
Comment (ClSlaid) [inline planner/core/logical_plan_test.go:1313]: You forgot to check `err` here.
Try run `make check` before every commit.

Comment (JmPotato) [reply planner/core/logical_plan_test.go:1313]: Actually, `errcheck` will ignore test files.

**Pull request merged**
